dalm/pipelines/reading_comprehension_pipeline.py

```diff
@@ -0,0 +1,382 @@
+    # Attention has been paid to try to save intermediate steps in case of failure
+    # so that the generation can be resumed from the last checkpoint
+    if comprehension_type in [SynthMode.LLM, SynthMode.BOTH] and llm_kwargs:
+        if generation_state_file:
+            if os.path.exists(generation_state_file):
+                with open(generation_state_file, "rb") as f:
+                    generation_state = pickle.load(f)
+            else:
+                generation_state = {"processed_files": [], "total_files": 0, "files_missed": 0}
+                pickle.dump(generation_state, open(generation_state_file, "wb"))
+
+        if not os.path.exists(llm_kwargs.dataset_output_path):
+            os.makedirs(llm_kwargs.dataset_output_path)
+
+        llm_rc_dataset_generator = generate_synthetic_dataset(
+            model_name=llm_kwargs.model_name,
+            input_directory_or_file=input,
+            processed_files=generation_state["processed_files"],
+            chunk=llm_kwargs.chunk or False,
+            context_length=llm_kwargs.context_length or 0,
+            csv_column=csv_column,
+        )
+
+        # generate llm based reading comprehension dataset
+        for index, filename, context, gen_text in llm_rc_dataset_generator:
+            qanda = question_and_answer_extractor(gen_text, context)
+            if qanda:
+                output_file = f"{filename}_{index}.json"
+                with open(os.path.join(llm_kwargs.dataset_output_path, output_file), "w") as o:
+                    json.dump(qanda, o)
+            else:
+                logger.warning(
+                    (
+                        f"No question and answer pairs found for {filename} " f"chunk: {index}"
+                        if llm_kwargs.chunk
+                        else ""
+                    )
+                )
+                generation_state["files_missed"] += 1
+            generation_state["processed_files"].append(filename)
+            generation_state["total_files"] += 1
+            pickle.dump(generation_state, open(generation_state_file, "wb"))
+
+        logger.info(" Statistics ")
+        success_files_count = generation_state["total_files"] - generation_state["files_missed"]
+        logger.info(f"Total number of successfully extracted q&a: {success_files_count}")
+        logger.info(f"Total files missed: {generation_state['files_missed']} out of {generation_state['total_files']}")
+        logger.info(f"Total files processed: {generation_state['total_files']}")
+
+        for file in os.listdir(llm_kwargs.dataset_output_path):
+            with open(os.path.join(llm_kwargs.dataset_output_path, file), "r") as f:
+                in_memory_dataset.append({"messages": json.load(f)})
+
+    # shuffle in memory dataset
+    random.shuffle(in_memory_dataset)
+
+    dataset = datasets.Dataset.from_list(in_memory_dataset)
+
+    if not os.path.exists(output_dataset_name):
+        dataset.save_to_disk(output_dataset_name)
```

Comment:
This is a bit surprising behavior.  If the caller passes in an empty directory that exists on disk, instead of saving the generated dataset to that directory, it will discard the dataset (not save it) and then later fail with an error:

```
FileNotFoundError: Directory output_dataset is neither a `Dataset` directory nor a `DatasetDict` directory.
```

A few ideas on how to fix:

1. Just overwrite the existing dataset with the new dataset (remove the `if not os.path.exists()` check)
2. Instead of asking the user to pass the name of the dataset, just generate a unique dataset directory and use that, then inform the user where the dataset was generated.